[Pipelining] fix extra memory usage in zero bubble

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #138720
* #138735
* #138504
* __->__ #138119

Full debugging details in here: https://docs.google.com/document/d/1Pe_E0KWAfsJ6MCvKZ5aR28rTXX-rYLg13XxwXd6AALw/edit?usp=sharing

In zero bubble, we have two methods `stage_backward_input` and `stage_backward_weight`. During `stage_backward_input` we compute the gradients of the input with respect to the stage outputs and also retain the graph of the autograd graph (different than 1F1B where `retain_graph=False`). The output / loss was still being retained across the next schedule step() because we return the loss to the user and use the output to the next step. To allow autograd to free the variables in the graph we need to detach the output/loss after we don't need to use it autograd anymore.

Pre-fix:
<img width="1021" alt="image" src="https://github.com/user-attachments/assets/6c8bf469-32b1-4dac-85ff-b97991f9f0e3">

Post-fix:
<img width="1039" alt="image" src="https://github.com/user-attachments/assets/a1875038-e80b-4dd4-84f2-38727d7792dc">

without AC (7B model on titan):
10% memory improvement

with AC (7B model on titan)
50% memory improvement


cc @XilunWu @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

H-Huang committed:
[PP] fix extra memory usage in zero bubble

[ghstack-poisoned]

wconstab commented on torch/distributed/pipelining/_backward.py:
```diff
@@ -201,6 +201,10 @@ def hook(grad_inputs):
             else:
                 inp.grad += dinputs[i]
+
+        for t in stage_outputs:
```
i would love a more detailed comment here about why this helps and why it is safe to do at this point

wconstab commented:
I'm wondering what we can do as far as unit testing for memory.

What if we put schedules aside and really focus on unit-level testing- We have a stage with a simple model, we know roughly how much activation memory is needed, and we set up hooks on cuda caching allocator to see if we allocate/free memory as expected after calling stage._forward / dI / dW.  And we do this with / without AC?

awgu commented:
I think you may not even need to hook into the caching allocator. Here is how we did unit testing for memory for FSDP2:
https://github.com/pytorch/pytorch/blob/443472b1cabac7758e59af7d80c25f99f3ec0fbd/test/distributed/_composable/fsdp/test_fully_shard_memory.py#L37

wconstab commented:
> To allow autograd to free the variables in the graph we need to detach the output/loss after we don't need to use it autograd anymore.

i would probably benefit from a graph diagram showing what is changed by setting outputs to detach=True.  

taking a stab, i think 'outputs' is the place we start the .grad computation from when we first do dInput calculation.  Then after dInput, we have to save some intermetiate nodes, but we shouldn't need to save outputs, and we also shouldn't need to save any other intermediate nodes that aren't needed for dW.  Then when we do dW calculation, we do not start from outputs, we start from saved intermediates.  

Is that right?  and does detaching the outputs free just the output node, or it also frees other parts of the autograd graph that are not needed for dW?

Also, how does loss fit in?  You didn't change 'loss' in this PR but I think you mentioned that we might need to clear loss tensor after printing it.  Can you explain more about that?

H-Huang committed:
Update on "[Pipelining] fix extra memory usage in zero bubble"


Full debugging details in here: https://docs.google.com/document/d/1Pe_E0KWAfsJ6MCvKZ5aR28rTXX-rYLg13XxwXd6AALw/edit?usp=sharing

In zero bubble, we have two methods `stage_backward_input` and `stage_backward_weight`. During `stage_backward_input` we compute the gradients of the input with respect to the stage outputs and also retain the graph of the autograd graph (different than 1F1B where `retain_graph=False`). The output / loss was still being retained across the next schedule step() because we return the loss to the user and use the output to the next step. To allow autograd to free the variables in the graph we need to detach the output/loss after we don't need to use it autograd anymore.

Pre-fix:
<img width="1021" alt="image" src="https://github.com/user-attachments/assets/6c8bf469-32b1-4dac-85ff-b97991f9f0e3">

Post-fix:
<img width="1039" alt="image" src="https://github.com/user-attachments/assets/a1875038-e80b-4dd4-84f2-38727d7792dc">

without AC (7B model on titan):
10% memory improvement

with AC (7B model on titan)
50% memory improvement


cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[Pipelining] fix extra memory usage in zero bubble"


Full debugging details in here: https://docs.google.com/document/d/1Pe_E0KWAfsJ6MCvKZ5aR28rTXX-rYLg13XxwXd6AALw/edit?usp=sharing

In zero bubble, we have two methods `stage_backward_input` and `stage_backward_weight`. During `stage_backward_input` we compute the gradients of the input with respect to the stage outputs and also retain the graph of the autograd graph (different than 1F1B where `retain_graph=False`). The output / loss was still being retained across the next schedule step() because we return the loss to the user and use the output to the next step. To allow autograd to free the variables in the graph we need to detach the output/loss after we don't need to use it autograd anymore.

Pre-fix:
<img width="1021" alt="image" src="https://github.com/user-attachments/assets/6c8bf469-32b1-4dac-85ff-b97991f9f0e3">

Post-fix:
<img width="1039" alt="image" src="https://github.com/user-attachments/assets/a1875038-e80b-4dd4-84f2-38727d7792dc">

without AC (7B model on titan):
10% memory improvement

with AC (7B model on titan)
50% memory improvement


cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang commented:
> Then after dInput, we have to save some intermetiate nodes, but we shouldn't need to save outputs, and we also shouldn't need to save any other intermediate nodes that aren't needed for dW. Then when we do dW calculation, we do not start from outputs, we start from saved intermediates.

Yep! This is all correct.

> does detaching the outputs free just the output node, or it also frees other parts of the autograd graph that are not needed for dW?

It frees the other parts of the autograd graph as well

> Also, how does loss fit in?

For the last stage, `loss` is passed in instead of the `stage_outputs` on which we call `autograd.grad()` and so this change will also detach the loss tensor since we don't need it anymore after the first backward. I was investigating why the last stage still uses more memory and I think I found the issue in the follow up PR

kwen2501 commented:
> > Then after dInput, we have to save some intermetiate nodes, but we shouldn't need to save outputs, and we also shouldn't need to save any other intermediate nodes that aren't needed for dW. Then when we do dW calculation, we do not start from outputs, we start from saved intermediates.
> 
> Yep! This is all correct.
> 
> > does detaching the outputs free just the output node, or it also frees other parts of the autograd graph that are not needed for dW?
> 
> It frees the other parts of the autograd graph as well
> 
> > Also, how does loss fit in?
> 
> For the last stage, `loss` is passed in instead of the `stage_outputs` on which we call `autograd.grad()` and so this change will also detach the loss tensor since we don't need it anymore after the first backward. I was investigating why the last stage still uses more memory and I think I found the issue in the follow up PR

This Q&A is great! 
I wonder if we could put it as comment in the code? Thanks!
